lib/streamlit/runtime/session_manager.py

```diff
@@ -0,0 +1,344 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import abstractmethod
+from dataclasses import dataclass
+from typing import Dict, List, Optional, TYPE_CHECKING
+
+from typing_extensions import Protocol
+
+from streamlit.proto.ForwardMsg_pb2 import ForwardMsg
+from .app_session import AppSession
+
+if TYPE_CHECKING:
+    from .runtime import Runtime
+
+
+class SessionClientDisconnectedError(Exception):
+    """Raised by operations on a disconnected SessionClient."""
+
+
+class SessionClient(Protocol):
+    """Interface for sending data to a session's client."""
+
+    @abstractmethod
+    def write_forward_msg(self, msg: ForwardMsg) -> None:
+        """Deliver a ForwardMsg to the client.
+
+        If the SessionClient has been disconnected, it should raise a
+        SessionClientDisconnectedError.
+        """
+        raise NotImplementedError
+
+
+# TODO(vdonato): We may want to consider either coming up with a better name for this or
+# consolidating it with the SessionData class since the two names are similar enough
+# to be confusing right now.
```

Comment:
I'm definitely in favor of this, if it makes design sense. (I also think `SessionData` is misnamed, because it's not just a data container! But also it contains junk that is definitely not session-related - script_path, script_folder, command_line - I don't think any of these vary based on session?)

Reply:
👍

I'll see if an obvious way of reworking these things comes to mind before this feature branch gets merged into `develop`, but I don't think it's terribly urgent to fix this now